domain/status/state/modelstate.go

```diff
@@ -1746,3 +1748,302 @@ GROUP BY u.application_uuid, application.name;
+	}
+
+	// Convert the internal status id from the
+	// (machine_cloud_instance_status_value table) into the core status.Status
+	// type.
+	machineStatus, err := status.DecodeCloudInstanceStatus(mStatus.Status)
+	if err != nil {
+		return status.StatusInfo[status.InstanceStatusType]{}, errors.Errorf("decoding cloud instance status for machine %q: %w", mName, err)
+	}
+
+	var since *time.Time
+	if mStatus.Updated.Valid {
+		since = &mStatus.Updated.V
+	}
+
+	return status.StatusInfo[status.InstanceStatusType]{
+		Status:  machineStatus,
+		Message: mStatus.Message,
+		Since:   since,
+		Data:    mStatus.Data,
+	}, nil
+}
+
+// SetInstanceStatus sets the cloud specific instance status for this
+// machine.
+// This method may return the following errors:
+// - [machineerrors.MachineNotFound] if the machine does not exist.
+// - [machineerrors.NotProvisioned] if the machine instance does not exist.
+func (st *ModelState) SetInstanceStatus(ctx context.Context, mName coremachine.Name, newStatus status.StatusInfo[status.InstanceStatusType]) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	statusID, err := status.EncodeCloudInstanceStatus(newStatus.Status)
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	nameIdent := machineName{Name: mName}
+	queryMachine := `SELECT uuid AS &machineUUID.* FROM machine WHERE name = $machineName.name;`
+	queryMachineStmt, err := st.Prepare(queryMachine, nameIdent, machineUUID{})
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	statusQuery := `
+INSERT INTO machine_cloud_instance_status (*)
+VALUES ($setMachineStatus.*)
+  ON CONFLICT (machine_uuid)
+  DO UPDATE SET 
+    status_id = excluded.status_id, 
+    message = excluded.message, 
+    updated_at = excluded.updated_at,
+    data = excluded.data;
+`
+	statusQueryStmt, err := st.Prepare(statusQuery, setMachineStatus{})
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	return db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		var mUUID machineUUID
+		if err := tx.Query(ctx, queryMachineStmt, nameIdent).Get(&mUUID); errors.Is(err, sqlair.ErrNoRows) {
+			return errors.Errorf("machine %q: %w", mName, machineerrors.MachineNotFound)
+		} else if err != nil {
+			return errors.Errorf("querying uuid for machine %q: %w", mName, err)
+		}
+
+		_, err := st.getInstanceID(ctx, tx, mUUID)
+		if err != nil {
+			return errors.Errorf("getting machine instance id for %q: %w", mName, err)
+		}
+
+		// Query for setting the machine cloud instance status
+		err = tx.Query(ctx, statusQueryStmt, setMachineStatus{
+			MachineUUID: mUUID.UUID,
+			StatusID:    statusID,
+			Message:     newStatus.Message,
+			Data:        newStatus.Data,
+			UpdatedAt:   newStatus.Since,
+		}).Run()
+		if err != nil {
+			return errors.Errorf("setting machine status for machine %q: %w", mName, err)
+		}
+		return nil
+	})
+}
+
+func (st *ModelState) getInstanceID(ctx context.Context, tx *sqlair.TX, mUUID machineUUID) (string, error) {
```

Comment:
mUUID should be a string.

Reply:
Not this one - `machineUUID` is a state type